### PR TITLE
Add PTX single/dual/triple wall switch support

### PR DESCRIPTION
**Pull request by itpansic**
[Product selling page](https://item.taobao.com/item.htm?id=595682430270&ali_refid=a3_430582_1006:1196370024:N:sTQ20SDXMrHIAJhaViK1vA%3D%3D:7087a66d02955d33a810ae5bab248853&ali_trackid=1_7087a66d02955d33a810ae5bab248853&spm=a230r.1.14.1#detail)
I added support with  3 style of this product (single, dual, triple), but only tested with the triple one.

So I hope someone who has PTX products can test and complete this script.

### CODE REVIEW

**Comment by syssi** [comment]
Please fix the lint issue!

**Comment by itpansic** [comment]
Am I doing it right? 
What should I do next? @syssi

File: miio/ptxswitch.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+            self.model = MODEL_PTX_TRIPLE_WALL_SWITCH
+
+    @command(
+        default_output=format_output(
+            "",
+            "Switch 1 Status: {result.is_on_1}\n"
+            "Switch 2 Status: {result.is_on_1}\n"
+            "Switch 2 Status: {result.is_on_1}\n"
+            "Switch 1 Name: {result.switch_name_1}\n"
+            "Switch 2 Name: {result.switch_name_1}\n"
```
Comment (syssi) [inline miio/ptxswitch.py]: Should be `result.switch_name_2`. Same applies to `is_on_1`.

File: miio/ptxswitch.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            "Switch 3 Name: {result.switch_name_1}\n")
+    )
+    def status(self) -> PtxSwitchStatus:
+        # Retrieve properties.
+        properties = AVAILABLE_PROPERTIES[self.model].copy()
+        param_list_1 = list()
```
Comment (syssi) [inline miio/ptxswitch.py]: Could you provide some example payload here as docstring? Could you provide better names for `param_list_1` and `param_list_2`? What do you split here?

File: miio/ptxswitch.py (hunk 1 above)
Comment (itpansic) [reply miio/ptxswitch.py]: Sorry... now fixed

File: miio/ptxswitch.py (hunk 2 above)
Comment (itpansic) [reply miio/ptxswitch.py]: I dumped something and added into miio/ptxswitch.py.
Querying the power status require params like [0,0,0].
Please see detail in dumped data.

**Comment by honcheng** [comment]
Hi, how can this be integrated in Home-Assistant?

File: miio/ptxswitch.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+             self.switch_name_2,
+             self.switch_name_3)
+        return s
+
+    def __json__(self):
+        return self.data
+
+
+class PtxSwitch(Device):
+    # Main class representing the PTX switch
```
Comment (rytilahti) [inline miio/ptxswitch.py]: docstrings need to be in tripple quotes: ```"""Main class representing the PTX switch."""

File: miio/ptxswitch.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+    def __init__(self, ip: str = None, token: str = None, start_id: int = 0,
+                 debug: int = 0, lazy_discover: bool = True,
+                 model: str = MODEL_PTX_TRIPLE_WALL_SWITCH) -> None:
+        super().__init__(ip, token, start_id, debug, lazy_discover)
+
+        if model in AVAILABLE_PROPERTIES:
+            self.model = model
+        else:
+            self.model = MODEL_PTX_TRIPLE_WALL_SWITCH
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Please add a logger warning for this fallback case.

File: miio/ptxswitch.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        # params for querying status of the switches
+        params_switch_status = list()
+
+        # params for querying other properties
+        params_other = list()
+
+        # Query every switch status
+
+        for k in properties:
+            if k[:5] == "is_on":
```
Comment (rytilahti) [inline miio/ptxswitch.py]: `if k.startswith()` is nicer form.

File: miio/ptxswitch.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        # (eg. 'switchname1') as the request param.
+
+        # params for querying status of the switches
+        params_switch_status = list()
+
+        # params for querying other properties
+        params_other = list()
+
+        # Query every switch status
+
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Remove this unnecessary whitespace.

File: miio/ptxswitch.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+            _LOGGER.debug(
+                "Count (%s) of requested params does not match the "
+                "count (%s) of received values.",
+                param_count, values_count)
+
+        # Query other params
+        # A single request is limited to 1 properties. Therefore the
+        # properties are divided into multiple requests
+
+        result_2 = list()
```
Comment (rytilahti) [inline miio/ptxswitch.py]: `res_other` for this and `res_status` for the `result_1` maybe?

File: miio/ptxswitch.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+            )
+            if len(value) >= 1:
+                v = value[0]
+                if param[:10] == "switchname":
+                    if isinstance(v, str):
+                        result_2.append(v)
+                    else:
+                        _LOGGER.debug("Unexpected type of switchname")
+                        result_2.append(None)
+
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Unnecessary whitespace.

File: miio/ptxswitch.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+        result_2 = list()
+        for param in params_other:
+            value = self.send(
+                "get_prop",
+                [param]
+            )
+            if len(value) >= 1:
+                v = value[0]
+                if param[:10] == "switchname":
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Use `startswith`.

File: miio/ptxswitch.py [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+            else:
+                result_2.append(None)
+                _LOGGER.debug("Property %s returns None.", param)
+
+        return PtxSwitchStatus(
+            defaultdict(lambda: None,
+                        zip(properties,
+                            result_1 + result_2)))
+
+    def turn_switch(self, index: int, switch_state: int) -> bool:
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Please use bool instead of int for the state, you can always convert between those using `bool(int_value)` and `int(bool_value)`.

File: miio/ptxswitch.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+            "SetSwitch{}".format(index),
+            [switch_state]
+        )
+        if result[:1] == [0] or result[:1] == [1]:
+            return True
+        else:
+            _LOGGER.debug("Toogle switch {} failed.".format(index))
+            return False
+
+    def turn_on_switch(self, index: int) -> bool:
```
Comment (rytilahti) [inline miio/ptxswitch.py]: This requires also `@command` decorator to make it visible to the cli.

File: miio/ptxswitch.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        return self.turn_switch(index, 1)
+
+    def turn_off_switch(self, index: int) -> bool:
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Please add `@command`.

File: miio/ptxswitch.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+        return self.turn_switch(index, 0)
+
+    def set_switch_name(self, index: int, name: str) -> bool:
```
Comment (rytilahti) [inline miio/ptxswitch.py]: Please add `@command` decorator.
